File: Go/K09.go
package main

import (
	"fmt"
	"sort"
	"strconv"
)

type K09 struct {}

func (k09 K09) execute(num int) {
	switch num {
	case 1: k09.question1()
	case 2: k09.question2()
	case 3: k09.question3()
	case 4: k09.question4()
	case 5: k09.question5()
	}
}

func (k09 K09) question1() {
	var ages [3]int
	var inputString string
	for i := 0; i < 3; i++ {
		fmt.Print(fmt.Sprintf("%d人目の年齢を入力＞", i + 1))
		_, _ = fmt.Scanln(&inputString)
		ages[i], _ = strconv.Atoi(inputString)
	}

	for i := 0; i < 25; i++ {
		fmt.Print("-")
	}
	fmt.Println()

	var average float32
	for i := 0; i < 3; i++ {
		fmt.Println(fmt.Sprintf("%d人目：%d歳", i + 1, ages[i]))
		average += float32(ages[i])
	}
	fmt.Println(fmt.Sprintf("平均年齢：%.1f", average / 3.0))
}

func (k09 K09) question2() {
	numbers := [5]int{8, 3, 12, 7, 9}
	fmt.Print("元々の配列：")
	for i := 0; i < len(numbers); i++ {
		fmt.Print(fmt.Sprintf("%d ", numbers[i]))
	}
	fmt.Println()

	fmt.Print("逆順での表示：")
	for i := len(numbers) - 1; i >= 0; i-- {
		fmt.Print(fmt.Sprintf("%d ", numbers[i]))
	}
}

func (k09 K09) question3() {
	studentScore := [3][4]int{
		{52, 71, 61, 47},
		{6, 84, 81, 20},
		{73, 98, 94, 95},
	}
	fmt.Println("\t\t|\t科目A\t科目B\t科目C\t科目D")
	for i := 0; i < 45; i++ {
		fmt.Print("-")
	}
	fmt.Println()

	for i := 0; i < len(studentScore); i++ {
		fmt.Print(fmt.Sprintf("学生%d\t|\t", i + 1))
		for j := 0; j < len(studentScore[i]); j++ {
			fmt.Print(fmt.Sprintf("%d\t\t", studentScore[i][j]))
		}
		fmt.Println()
	}
}

func (k09 K09) question4() {
	studentScore := [3][4]int{
		{52, 71, 61, 47},
		{6, 84, 81, 20},
		{73, 98, 94, 95},
	}

	total := 0
	var average = [4]float32{0.0, 0.0, 0.0, 0.0}

	fmt.Println("\t\t|\t科目A\t科目B\t科目C\t科目D\t|\t合計点")
	for i := 0; i < 65; i++ {
		fmt.Print("-")
	}
	fmt.Println()

	for i := 0; i < len(studentScore); i++ {
		fmt.Print(fmt.Sprintf("学生%d\t|\t", i + 1))
		for j := 0; j < len(studentScore[i]); j++ {
			fmt.Print(fmt.Sprintf("%d\t\t", studentScore[i][j]))
			total += studentScore[i][j]
			average[j] += float32(studentScore[i][j])
		}
		fmt.Print(fmt.Sprintf("|\t%d", total))
		total = 0
		fmt.Println()
	}
	fmt.Print("平均点\t|\t")
	for i := 0; i < len(average); i++ {
		fmt.Print(fmt.Sprintf("%.1f\t", average[i] / 3.0))
	}
	fmt.Println("|")
}

func (k09 K09) question5() {
	count := 0
	var numbers []int
	var inputString string

	for i := 0; i < 100; i++ {
		fmt.Print(fmt.Sprintf("%d件目の入力：", i + 1))
		_, _ = fmt.Scanln(&inputString)
		result, _ := strconv.Atoi(inputString)
		numbers = append(numbers, result)
		if numbers[count] < 0	 {
			break
		}
		count++
	}
	fmt.Println("----並び替え後----")
	numSlice := numbers[:count]
	sort.Ints(numSlice)

	for i := 0; i < len(numSlice); i++ {
		fmt.Println(fmt.Sprintf("%d", numSlice[i]))
	}
}

File: rust/src/k09.rs
use crate::Executable;
use crate::read_user_input::read_user_input;
use std::io::Write;

pub struct K09 {}

impl K09 {
    pub fn new() -> K09 {
        K09 {}
    }

    pub fn question_5(&self) {
        let mut count = 0;
        let mut numbers: [i32; 100] = [0; 100];
        for x in 0..numbers.len() {
            print!("{}件目の入力：", x + 1);
            std::io::stdout().flush().unwrap();
            let mut input_string = String::new();
            numbers[x] = match read_user_input(&mut input_string) {
                Ok(x) => x,
                Err(e) => {
                    eprintln!("エラーが発生しました：{}", e.to_string());
                    return;
                }
            };
            if numbers[count] < 0 {
                break;
            }
            count += 1;
        }

        println!("----並び替え後----");
        let mut num_slice = &mut numbers[..count];
        num_slice.sort();

        for x in num_slice.iter() {
            println!("{}", x);
        }
    }

}

impl Executable for K09 {

    fn execute(&self, chapter: u32)
    {
        match chapter
            {
                1 => self.question_1(),
                2 => self.question_2(),
                3 => self.question_3(),
                4 => self.question_4(),
                5 => self.question_5(),
                _ => ()
            }
    }

    fn question_1(&self) {
        let mut ages: [i32; 3] = [0; 3];
        for i in 0..ages.len() {
            print!("{}人目の年齢を入力＞", i + 1);
            std::io::stdout().flush().unwrap();
            let mut input_str = String::new();
            ages[i] = match read_user_input(&mut input_str) {
                Ok(x) => x,
                Err(e) => {
                    eprintln!("エラーが発生しました：{}", e.to_string());
                    return;
                }
            }
        }

        for _ in 0..20 {
            print!("-");
        }
        println!();

        for age in ages.iter().enumerate() {
            println!("{}人目：{}歳", age.0 as i32 + 1, age.1);
        }
        let average: f32 = ages.iter().map(|x| {
            *x as f32
        }).sum::<f32>() / ages.len() as f32;
        println!("平均年齢：{:.1}", average);
    }

    fn question_2(&self) {
        let numbers: [i32; 5] = [8, 3, 12, 7, 9];
        print!("元々の配列：");
        for x in numbers.iter() {
            print!("{} ", x);
        }
        println!();

        print!("逆順での表示：");
        for x in numbers.iter().rev() {
            print!("{} ", x);
        }
        std::io::stdout().flush().unwrap();
    }

    fn question_3(&self) {
        let student_scores: [[i32; 4]; 3] = [
            [52, 71, 61, 47],
            [6, 84, 81, 20],
            [73, 98, 94, 95]
        ];

        println!("\t\t|\t科目A\t科目B\t科目C\t科目D");
        for _ in 0..45 {
            print!("-");
        }
        println!();

        for student in student_scores.iter().enumerate() {
            print!("学生{}\t|\t", student.0 as i32 + 1);
            for score in student_scores[student.0].iter() {
                print!("{}\t\t", score);
            }
            println!();
        }
    }

    fn question_4(&self) {
        let student_scores: [[i32; 4]; 3] = [
            [52, 71, 61, 47],
            [6, 84, 81, 20],
            [73, 98, 94, 95]
        ];

        let mut total: i32 = 0;
        let mut average: [f32; 4] = [0.0; 4];

        println!("\t\t|\t科目A\t科目B\t科目C\t科目D\t|\t合計点");
        for _ in 0..65 {
            print!("-");
        }
        println!();

        for student in student_scores.iter().enumerate() {
            print!("学生{}\t|\t", student.0 as i32 + 1);
            for score in student_scores[student.0].iter().enumerate() {
                print!("{}\t\t", score.1);
                total += score.1;
                average[score.0] += *score.1 as f32;
            }
            print!("|\t{}", total);
            total = 0;
            println!();
        }
        print!("平均点\t|\t");
        for x in average.iter() {
            print!("{:.1}\t", x / 3.0);
        }
        println!("|");
    }

}